OCD-4050: Don't allow all products to be transferred from developer

kekey1 opened the pull request.

kekey1 committed:
fix: Handle null certification date due to bad activity record

[#OCD-4050]

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
fix!: Support blank contact fields supplied during product edit

[#OCD-3629]

kekey1 committed:
fix!: Ensure a developer cannot have 0 products

[#OCD-4050]

kekey1 committed:
refactor: Update error message text to tell user to contact ONC

[#OCD-4050]

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
Merge remote-tracking branch 'upstream/staging' into OCD-4050

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
fix!: Product ownership change should correctly record listing activity

Product ownership changes were getting both the "before" and "after"
listing data from the shared store so the "after" product did not
include product or developer changes.

[#OCD-4050]

kekey1 committed:
fix: Validate that developer has other products when transfer w/o hist

There was validation added to check that the old developer product owner
has other products when product ownership is changed with an entry in
the history. However, if there is no entry in the history (user clicks
'x' in the UI to remove the automatically added history entry) then we
were allowing the ownership change regardless. This commit adds code to
validate the "developer must have at least one other product" rule in
the event of an ownership-without-history-change.

[#OCD-4050]

kekey1 committed:
fix: Rollback on developer merge failure for any Exception

If a developer merge failed after the new developer was created, the new
developer was hanging around in the system. After altering the
annotation to rollback for any Exception that newly created developer is
also removed.

[OCD-4050]

kekey1 committed:
fix: Allow developer to lose all products during developer merge

We created a rule that developers could not lose all of their products.
However, applying this rule on developer merge makes developer merge not
work (because by definition we are removing all the products of a
developer!).

I don't love the solution I came up with here but I thought about it for
a while and did not see a better way. I am settling for it since we
likely will be re-writing this workflow relatively soon with a ticket in
the backlog that is going to let developers transfer ownership of ALL
products from one to another.

[#OCD-4050]

kekey1 committed:
feat: Update developer merge email to standard CHPL HTML format

[#OCD-4050]

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
feat: Ignore files being generated by Eclipse

[#OCD-4050]

kekey1 committed:
fix: Use Dynamic Update HQL on product/owner entities

The soft delete triggers combined with non-dymanic HQL updates were
causing previously deleted product owner history entries to come back if
those products were part of a developer merge.

The dynamic update annotation only puts the fields in the update
statement if they were changed.

[#OCD-4050]

kekey1 committed:
fix: Do not consider "deleted" developers for "must have a product" rule

When merging developers, the flow is to transfer all products from both
developers to a newly created developer. Then mark the merged developers
as deleted. If a person later goes to edit ownership history of one of
the moved products and edits it in a certain way, they get an error
related to the "deleted" developer. This commit adds code to not care
about whether deleted developers have any products or not.

[#OCD-4050]

kekey1 committed:
fix!: Handle null developer contact name in Developer Access Report

We got into the state where a developer had null contact info but was
still included in the access report when we changed some product
ownership. If a product that has an active listing is transferred to a
developer that only has retired listings, then that developer now must
be included in the report. The developer previously was never required
to have contact information and we did not require it when transferring
the product. Maybe we should?

[#OCD-4050]

andlar commented on .gitignore:
```diff
@@ -16,6 +16,8 @@ local.properties
 .loadpath
 target/
+chpl/chpl-api/src/main/webapp/META-INF/
+chpl/chpl-api/src/main/webapp/.gitignore
```
Usually the `.gitignore` file is saved in git, to make sure all the things git should ignore get ignored. Is there something different about this file?

kekey1 replied:
Ugh. My new version of Eclipse is auto-generating this META-INF directory AND auto-generating a .gitignore inside of it. I think it's a setting that I can't figure out... I have tried. Todd doesn't have this problem but all of our settings look the same. I can get Eclipse to STOP auto-generating the folder/file but then I'm not able to use the built-in Tomcat and deploy the project as a web application. I have been trying so hard not to commit these auto-generated files but I'm sure I'll mess up at some point... so I decided that it was relatively harmless to add the Eclipse auto-generated stuff to the main .gitignore and that's what I did.

andlar replied:
Ignoring the autogenerated stuff is a great idea, but this change also ignores the file that determines what stuff is ignored, I think. I can't think of a reason to ignore the file that defines what should be ignored. I'm fine with line 19, but is there stuff in the file in line 20 that now won't be ignored because it we're ignoring the file that tells git to ignore it?

kekey1 committed:
refactor: Undo ignoring of autogenerated ignore file

[#OCD-4050]

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
Merge branch 'staging' into OCD-4050

kekey1 committed:
fix: Remove links from merged developers

The links in the email all resulted in 404s. This has been existing for
some time.

[#OCD-4050]
